File: {{cookiecutter.project_folder}}/tutorial/eigen.cpp
// https://github.com/wjakob/nanobind/blob/master/tests/test_eigen.cpp
// https://github.com/wjakob/nanobind/blob/master/tests/test_eigen.py
#include "compas.h"
#include <nanobind/eigen/dense.h>

/**
 * Type aliases for cleaner code and better readability
 * These types simplify working with Eigen matrices/vectors with specific memory layouts
 */
// Matrix types
using ColMatrixXf = Eigen::MatrixXf; // Default is column-major
using RowMatrixXf = Eigen::Matrix<float, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;

// Vector types
using VectorXf = Eigen::VectorXf;       // Column vector (Nx1)
using RowVectorXf = Eigen::RowVectorXf; // Row vector (1xN)

/**
 * Create a column-major matrix (Eigen default)
 * @param rows Number of rows in the matrix
 * @param cols Number of columns in the matrix
 * @return Column-major matrix filled with sequential values
 */
ColMatrixXf matrix_colmajor(int rows, int cols) {
    ColMatrixXf mat(rows, cols);
    int count = 0;
    for (int c = 0; c < cols; c++) {
        for (int r = 0; r < rows; r++) {
            mat(r, c) = ++count;
        }
    }
    return mat;
}

/**
 * Create a row-major matrix
 * @param rows Number of rows in the matrix
 * @param cols Number of columns in the matrix
 * @return Row-major matrix filled with sequential values
 */
RowMatrixXf matrix_rowmajor(int rows, int cols) {
    RowMatrixXf mat(rows, cols);
    int count = 0;
    for (int r = 0; r < rows; r++) {
        for (int c = 0; c < cols; c++) {
            mat(r, c) = ++count;
        }
    }
    return mat;
}

/**
 * Accept only column-major matrices
 * @param x Column-major matrix reference, will reject row-major matrices
 */
void matrix_colmajor_only(const Eigen::Ref<const ColMatrixXf>& x) {
    printf("Column-major matrix %ldx%ld\n", x.rows(), x.cols());
}

/**
 * Accept only row-major matrices
 * @param x Row-major matrix reference, will reject column-major matrices
 */
void matrix_rowmajor_only(const Eigen::Ref<const RowMatrixXf>& x) {
    printf("Row-major matrix %ldx%ld\n", x.rows(), x.cols());
}

/**
 * Zero-copy for any layout using DRef
 * @param x Matrix reference that accepts any memory layout without copying
 */
void matrix_dref(const nb::DRef<Eigen::MatrixXf>& x) {
    printf("Matrix %ldx%ld\n", x.rows(), x.cols());
}

/**
 * Modify a matrix element in-place (zero-copy)
 * @param x Non-const matrix reference that can be modified
 */
void matrix_modify(Eigen::Ref<ColMatrixXf> x) {
    // Correct syntax for modifying a matrix element
    x(0, 0) = 99.0f;
    printf("Modified matrix[0,0] = %.1f\n", x(0, 0));
}

/**
 * Safe sum with explicit return type for proper evaluation
 * @param a First matrix to sum
 * @param b Second matrix to sum
 * @return Result of matrix addition, properly evaluated for zero-copy return
 */
ColMatrixXf matrix_sum(const ColMatrixXf& a, const ColMatrixXf& b) {
    return a + b;
}

//====================== VECTOR OPERATIONS ======================//

/**
 * Create a vector and return it efficiently (zero-copy)
 * @param size Size of the vector to create
 * @return Column vector filled with sequential values, transferred to Python without copying
 */
VectorXf vector_create(int size) {
    VectorXf vec(size);
    for (int i = 0; i < size; i++) {
        vec(i) = i + 1.0f;
    }
    // Zero-copy when returned by value
    return vec;
}

/**
 * Zero-copy modify vector in-place (uses non-const reference)
 * @param v Vector reference that can be modified, changes reflect in Python
 */
void vector_modify(Eigen::Ref<VectorXf> v) {
    // Directly modifies the input vector
    if (v.size() > 0) {
        v(0) = 99.0f;
        printf("Modified vector[0] = %.1f\n", v(0));
    }
}

//====================== MAP OPERATIONS ======================//

/**
 * Map a 1D array to an Eigen vector and modify it in-place
 * @param array NumPy array to map to an Eigen vector
 * @throws std::runtime_error if the array is not 1D
 */
void map_vector(nb::ndarray<float> array) {
    if (array.ndim() != 1) {
        throw std::runtime_error("Expected a 1D array");
    }
    
    // Map NumPy array to Eigen vector (zero-copy)
    Eigen::Map<VectorXf> vec(array.data(), array.shape(0));
    
    // Modify the vector (changes reflect in NumPy array)
    vec *= 2.0f;
}

/**
 * Map a 2D array to an Eigen matrix and modify it in-place
 * @param array NumPy array to map to an Eigen matrix
 * @throws std::runtime_error if the array is not 2D
 */
void map_matrix(nb::ndarray<float> array) {
    if (array.ndim() != 2) {
        throw std::runtime_error("Expected a 2D array");
    }
    
    // Map NumPy array to Eigen matrix (zero-copy)
    Eigen::Map<Eigen::MatrixXf> mat(array.data(), array.shape(0), array.shape(1));
    
    // Modify the matrix (changes reflect in NumPy array)
    mat *= 2.0f;
}

NB_MODULE(_eigen, m) {
    m.doc() = "Eigen matrix and vector examples with efficient zero-copy operations";
    
    // Matrix functions
    m.def("colmajor", &matrix_colmajor, "Create a column-major matrix",
          nb::arg("rows") = 3, nb::arg("cols") = 4);
    m.def("rowmajor", &matrix_rowmajor, "Create a row-major matrix",
          nb::arg("rows") = 3, nb::arg("cols") = 4);
    m.def("colmajor_only", &matrix_colmajor_only, "Accept only column-major matrices");
    m.def("rowmajor_only", &matrix_rowmajor_only, "Accept only row-major matrices");
    m.def("dref", &matrix_dref, "Zero-copy for any layout using DRef");
    m.def("modify", &matrix_modify, "Modify a matrix element (non-const reference)");
    m.def("sum", &matrix_sum, "Sum two matrices");
    
    // Vector functions - minimal zero-copy examples
    m.def("vector", &vector_create, "Create a vector (zero-copy return)",
          nb::arg("size") = 5);
    m.def("vector_modify", &vector_modify, "Modify a vector in-place (zero-copy)");
    
    // Map functions - direct memory access examples
    m.def("map_vector", &map_vector, "Map 1D array to Eigen vector and modify in-place");
    m.def("map_matrix", &map_matrix, "Map 2D array to Eigen matrix and modify in-place");
}

File: {{cookiecutter.project_folder}}/tutorial/eigen.py
from {{cookiecutter.package_name}} import _eigen
import numpy as np

# Matrix examples
def create_colmajor_matrix():
    return _eigen.colmajor(3, 4)  # Column-major (Eigen default)

def create_rowmajor_matrix():
    return _eigen.rowmajor(3, 4)  # Row-major

def test_layout_specific(col_mat, row_mat):
    # Original matrices created in C++
    _eigen.colmajor_only(col_mat)  # Only accepts column-major
    _eigen.rowmajor_only(row_mat)  # Only accepts row-major

def test_layout_specific_numpy():
    # Create matrices directly with NumPy using different memory layouts
    np_col = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float32, order='F')  # Column-major layout
    np_row = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float32, order='C')  # Row-major layout (default)
    
    # Pass to layout-specific functions
    _eigen.colmajor_only(np_col)  # Works with column-major NumPy array
    _eigen.rowmajor_only(np_row)   # Works with row-major NumPy array
    
    return np_col, np_row

def modify_matrix():
    matrix = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float32, order='F')  # Column-major
    _eigen.modify(matrix)  # Modifies element [0,0] to 99
    return matrix

# Vector examples
def create_vector(size=5):
    return _eigen.vector(size)

def modify_vector():
    vec = np.array([1, 2, 3, 4, 5], dtype=np.float32)
    _eigen.vector_modify(vec)  # Modifies element [0] to 99
    return vec

# Eigen::Map examples
def map_vector_example():
    vector = np.array([1, 2, 3, 4, 5], dtype=np.float32)
    _eigen.map_vector(vector)  # Doubles all values
    return vector

def map_matrix_example():
    matrix = np.array([[1, 2, 3], [4, 5, 6]], dtype=np.float32)
    _eigen.map_matrix(matrix)  # Doubles all values
    return matrix

# Matrix examples using matrices created in C++
col_mat = create_colmajor_matrix()
row_mat = create_rowmajor_matrix()
test_layout_specific(col_mat, row_mat)

# Matrix examples using NumPy arrays directly
np_col, np_row = test_layout_specific_numpy()
modified_matrix = modify_matrix()

# Vector examples
vector = create_vector()
modified_vector = modify_vector()

# Map examples
mapped_vector = map_vector_example()
mapped_matrix = map_matrix_example()

# Print results
print(f"Modified matrix[0,0]: {modified_matrix[0,0]}")  # Should be 99
print(f"Modified vector[0]: {modified_vector[0]}")      # Should be 99
print(f"Mapped vector[0]: {mapped_vector[0]}")          # Should be 2 (doubled)
print(f"Mapped matrix[0,0]: {mapped_matrix[0,0]}")      # Should be 2 (doubled)
